packages/utils/telemetry-utils/src/sampledTelemetryHelper.ts

```diff
@@ -0,0 +1,131 @@
+
+    /**
+     * @param eventBase - Custom properties to include in the telemetry performance event when it is written.
+     * @param logger - The logger to use to write the telemetry performance event.
+     * @param includeAggregateMetrics - If set to `true`, the telemetry performance event will include aggregated
+     *                                  metrics (execution count, total duration, min duration, max duration) for
+     *                                  all the executions in between generated events.
+     */
+    public constructor(
+        private readonly eventBase: any,
+        private readonly logger: ITelemetryLogger,
+        private readonly includeAggregateMetrics: boolean = false) {
+    }
+
+    /**
+     * @param codeToMeasure - The code to be executed and measured.
+     * @param countThreshold - The telemetry performance event with aggregated data will be
+     *                         generated after this many executions of the code block.
+     * @param dimension - A key to track executions of the code block separately. Each different
+     *                    value of this parameter has a separate set of executions and metrics tracked
+     *                    by the class. If no such distinction needs to be made, do not provide a value.
+     * @returns Whatever the passed-in code block returns.
+     */
+    public measure<T>(
+        codeToMeasure: () => T,
+        countThreshold: number,
+        dimension: string = ""): T {
+        const start = performance.now();
+        const returnValue = codeToMeasure();
+        const duration = performance.now() - start;
+
+        let m = this.measurementsMap.get(dimension);
+        if (m === undefined) {
+            m = {
+                latestDuration: 0,
+                executionCount: 0,
+                totalDuration: 0,
+                minDuration: undefined,
+                maxDuration: undefined,
+            };
+            this.measurementsMap.set(dimension, m);
+        }
+        m.executionCount++;
+        m.latestDuration = duration;
+
+        if (this.includeAggregateMetrics) {
+            m.totalDuration += duration;
+            m.minDuration = m.minDuration === undefined
+                ? duration
+                : Math.min(m.minDuration, duration);
+            m.maxDuration = m.maxDuration === undefined
+                ? duration
+                : Math.max(m.maxDuration, duration);
+        }
+
+        if (m.executionCount >= countThreshold) {
+            this.flushDimension(dimension);
+        }
+
+        return returnValue;
+    }
+
+    private flushDimension(dimension: string) {
+        const measurements = this.measurementsMap.get(dimension);
+        if (measurements === undefined) {
+            return;
+        }
+
+        if (measurements.executionCount !== 0) {
+            let telemetryEvent: ITelemetryPerformanceEvent = {
+                ...this.eventBase,
+                ...{
```

Comment:
this is wierd, why create and object just to expand it, rather than just putting the properties on the object

Reply:
i also think you can build the whole object in one go
```
{
... this.eventBase,
... measurement,
dimension
}
```
we should avoid spurious object allocations when we don't need them

Reply:
just fix the names of the measurement object to match the event. it's a bit awkward to have them different

Reply:
Makes sense. I'll have to refactor a bit to keep an execution count outside of the measurement object, so it's not always populated there and stamped in the final event (which I think could be confusing if the user specified no aggregate properties).

Reply:
i think  it is fine to always log. it will also help with tuning the sample frequency.